File: Tweak/Modules/DarwinNotifyManager.m
//
//  DarwinNotifyManager.m
//  GhostKit Tweak
//

#import "DarwinNotifyManager.h"
#import <CoreFoundation/CoreFoundation.h>

static NSString *const kGhostKitDir      = @"/var/mobile/Library/GhostKit";
static NSString *const kCommandFilePath  = @"/var/mobile/Library/GhostKit/command.plist";
static NSString *const kResultFilePath   = @"/var/mobile/Library/GhostKit/result.plist";

@implementation DarwinNotifyManager

+ (instancetype)sharedInstance {
    static DarwinNotifyManager *instance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[DarwinNotifyManager alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        [[NSFileManager defaultManager] createDirectoryAtPath:kGhostKitDir
                      withIntermediateDirectories:YES
                                       attributes:nil
                                            error:nil];
    }
    return self;
}

#pragma mark - Send command (App → Tweak path)

- (nullable NSDictionary *)sendCommand:(DarwinCommand)command
                              params:(nullable NSDictionary *)params
                             timeout:(NSTimeInterval)timeout {
    // Remove stale result
    [[NSFileManager defaultManager] removeItemAtPath:kResultFilePath error:nil];

    // Write command
    NSMutableDictionary *cmd = [@{ @"command": @(command) } mutableCopy];
    if (params.count) cmd[@"params"] = params;
    [cmd writeToFile:kCommandFilePath atomically:YES];

    // Post Darwin notification
    NSString *notifName = [DarwinNotifyManager notificationNameForCommand:command];
    CFNotificationCenterRef center = CFNotificationCenterGetDarwinNotifyCenter();
    CFNotificationCenterPostNotification(center,
        (__bridge CFNotificationName)notifName, nil, nil, true);

    // Poll for result
    NSTimeInterval deadline = [[NSDate date] timeIntervalSinceReferenceDate] + timeout;
    while ([[NSDate date] timeIntervalSinceReferenceDate] < deadline) {
        NSDictionary *result = [NSDictionary dictionaryWithContentsOfFile:kResultFilePath];
        if (result) {
            NSNumber *rc = result[@"command"];
            if (rc && rc.integerValue == command) return result;
        }
        usleep(50000); // 50ms
    }
    NSLog(@"[GhostKit] Darwin notify timeout for command %ld", (long)command);
    return nil; // NSDictionary return type is OK
}

- (void)postNotification:(DarwinCommand)command params:(nullable NSDictionary *)params {
    NSMutableDictionary *cmd = [@{ @"command": @(command) } mutableCopy];
    if (params.count) cmd[@"params"] = params;
    [cmd writeToFile:kCommandFilePath atomically:YES];
    NSString *notifName = [DarwinNotifyManager notificationNameForCommand:command];
    CFNotificationCenterRef center = CFNotificationCenterGetDarwinNotifyCenter();
    CFNotificationCenterPostNotification(center,
        (__bridge CFNotificationName)notifName, nil, nil, true);
}

#pragma mark - Helpers

+ (DarwinCommand)commandFromString:(NSString *)name {
    if ([name isEqualToString:@"GhostKitCleanKeychain"])            return DarwinCommandCleanKeychain;
    if ([name isEqualToString:@"GhostKitDeepCleanKeychain"])        return DarwinCommandDeepCleanKeychain;
    if ([name isEqualToString:@"GhostKitDeleteAllKeychains"])       return DarwinCommandDeleteAllKeychains;
    if ([name isEqualToString:@"GhostKitBackupKeychain"])           return DarwinCommandBackupKeychain;
    if ([name isEqualToString:@"GhostKitRestoreKeychain"])          return DarwinCommandRestoreKeychain;
    if ([name isEqualToString:@"GhostKitListKeychain"])             return DarwinCommandListKeychain;
    if ([name isEqualToString:@"GhostKitRefreshIDFA"])              return DarwinCommandRefreshIDFA;
    if ([name isEqualToString:@"GhostKitChangeIdentifier"])         return DarwinCommandChangeIdentifier;
    if ([name isEqualToString:@"GhostKitGetCurrentIdentifiers"])    return DarwinCommandGetCurrentIdentifiers;
    if ([name isEqualToString:@"GhostKitResetDevice"])              return DarwinCommandResetDevice;
    if ([name isEqualToString:@"GhostKitGetDeviceInfo"])            return DarwinCommandGetDeviceInfo;
    if ([name isEqualToString:@"GhostKitCleanSystemResidue"])       return DarwinCommandCleanSystemResidue;
    if ([name isEqualToString:@"GhostKitCleanDatabaseCache"])       return DarwinCommandCleanDatabaseCache;
    if ([name isEqualToString:@"GhostKitCleanDataDirectory"])       return DarwinCommandCleanDataDirectory;
    if ([name isEqualToString:@"GhostKitCleanCookies"])             return DarwinCommandCleanCookies;
    if ([name isEqualToString:@"GhostKitCleanPasteboard"])          return DarwinCommandCleanPasteboard;
    if ([name isEqualToString:@"GhostKitGetAppSize"])               return DarwinCommandGetAppSize;
    if ([name isEqualToString:@"GhostKitGetAllApps"])               return DarwinCommandGetAllApps;
    if ([name isEqualToString:@"GhostKitSearchApps"])               return DarwinCommandSearchApps;
    if ([name isEqualToString:@"GhostKitUninstallApp"])             return DarwinCommandUninstallApp;
    if ([name isEqualToString:@"GhostKitGetAppInfo"])               return DarwinCommandGetAppInfo;
    if ([name isEqualToString:@"GhostKitGetAccountList"])           return DarwinCommandGetAccountList;
    if ([name isEqualToString:@"GhostKitAddAccount"])               return DarwinCommandAddAccount;
    if ([name isEqualToString:@"GhostKitDeleteAccount"])            return DarwinCommandDeleteAccount;
    if ([name isEqualToString:@"GhostKitGetCurrentAccount"])        return DarwinCommandGetCurrentAccount;
    if ([name isEqualToString:@"GhostKitSwitchAccount"])            return DarwinCommandSwitchAccount;
    if ([name isEqualToString:@"GhostKitAllowPasteForAll"])         return DarwinCommandAllowPasteForAll;
    if ([name isEqualToString:@"GhostKitIsPasteAllowed"])           return DarwinCommandIsPasteAllowed;
    if ([name isEqualToString:@"GhostKitSafeExit"])                 return DarwinCommandSafeExit;
    if ([name isEqualToString:@"GhostKitRespring"])                 return DarwinCommandRespring;
    if ([name isEqualToString:@"GhostKitLdrestart"])                return DarwinCommandLdrestart;
    if ([name isEqualToString:@"GhostKitInjectDylib"])              return DarwinCommandInjectDylib;
    if ([name isEqualToString:@"GhostKitRemoveDylib"])              return DarwinCommandRemoveDylib;
    if ([name isEqualToString:@"GhostKitGetInjectedDylibs"])        return DarwinCommandGetInjectedDylibs;
    if ([name isEqualToString:@"GhostKitApplyGraphicsConfig"])      return DarwinCommandApplyGraphicsConfig;
    if ([name isEqualToString:@"GhostKitGetAvailablePresets"])      return DarwinCommandGetAvailablePresets;
    if ([name isEqualToString:@"GhostKitGetCurrentGraphics"])       return DarwinCommandGetCurrentGraphics;
    if ([name isEqualToString:@"GhostKitRestoreDefaultGraphics"])   return DarwinCommandRestoreDefaultGraphics;
    if ([name isEqualToString:@"GhostKitShowSettings"])             return DarwinCommandShowSettings;
    return DarwinCommandUnknown;
}

+ (NSString *)notificationNameForCommand:(DarwinCommand)command {
    switch (command) {
        case DarwinCommandCleanKeychain:            return @"GhostKitCleanKeychain";
        case DarwinCommandDeepCleanKeychain:        return @"GhostKitDeepCleanKeychain";
        case DarwinCommandDeleteAllKeychains:       return @"GhostKitDeleteAllKeychains";
        case DarwinCommandBackupKeychain:           return @"GhostKitBackupKeychain";
        case DarwinCommandRestoreKeychain:          return @"GhostKitRestoreKeychain";
        case DarwinCommandListKeychain:             return @"GhostKitListKeychain";
        case DarwinCommandRefreshIDFA:              return @"GhostKitRefreshIDFA";
        case DarwinCommandChangeIdentifier:         return @"GhostKitChangeIdentifier";
        case DarwinCommandGetCurrentIdentifiers:    return @"GhostKitGetCurrentIdentifiers";
        case DarwinCommandResetDevice:              return @"GhostKitResetDevice";
        case DarwinCommandGetDeviceInfo:            return @"GhostKitGetDeviceInfo";
        case DarwinCommandCleanSystemResidue:       return @"GhostKitCleanSystemResidue";
        case DarwinCommandCleanDatabaseCache:       return @"GhostKitCleanDatabaseCache";
        case DarwinCommandCleanDataDirectory:       return @"GhostKitCleanDataDirectory";
        case DarwinCommandCleanCookies:             return @"GhostKitCleanCookies";
        case DarwinCommandCleanPasteboard:          return @"GhostKitCleanPasteboard";
        case DarwinCommandGetAppSize:               return @"GhostKitGetAppSize";
        case DarwinCommandGetAllApps:               return @"GhostKitGetAllApps";
        case DarwinCommandSearchApps:               return @"GhostKitSearchApps";
        case DarwinCommandUninstallApp:             return @"GhostKitUninstallApp";
        case DarwinCommandGetAppInfo:               return @"GhostKitGetAppInfo";
        case DarwinCommandGetAccountList:           return @"GhostKitGetAccountList";
        case DarwinCommandAddAccount:               return @"GhostKitAddAccount";
        case DarwinCommandDeleteAccount:            return @"GhostKitDeleteAccount";
        case DarwinCommandGetCurrentAccount:        return @"GhostKitGetCurrentAccount";
        case DarwinCommandSwitchAccount:            return @"GhostKitSwitchAccount";
        case DarwinCommandAllowPasteForAll:         return @"GhostKitAllowPasteForAll";
        case DarwinCommandIsPasteAllowed:           return @"GhostKitIsPasteAllowed";
        case DarwinCommandSafeExit:                 return @"GhostKitSafeExit";
        case DarwinCommandRespring:                 return @"GhostKitRespring";
        case DarwinCommandLdrestart:                return @"GhostKitLdrestart";
        case DarwinCommandInjectDylib:              return @"GhostKitInjectDylib";
        case DarwinCommandRemoveDylib:              return @"GhostKitRemoveDylib";
        case DarwinCommandGetInjectedDylibs:        return @"GhostKitGetInjectedDylibs";
        case DarwinCommandApplyGraphicsConfig:      return @"GhostKitApplyGraphicsConfig";
        case DarwinCommandGetAvailablePresets:      return @"GhostKitGetAvailablePresets";
        case DarwinCommandGetCurrentGraphics:       return @"GhostKitGetCurrentGraphics";
        case DarwinCommandRestoreDefaultGraphics:   return @"GhostKitRestoreDefaultGraphics";
        case DarwinCommandShowSettings:             return @"GhostKitShowSettings";
        default:                                    return @"GhostKitUnknown";
    }
}

@end

File: App/GhostKitApp/Managers/DarwinNotifyManager.swift
//
//  DarwinNotifyManager.swift
//  GhostKit
//
//  App-side Darwin notification manager.
//  Sends cross-process notifications to the Tweak layer (injected into
//  target apps).  Since Darwin notifications cannot carry userInfo,
//  command parameters are written to a shared plist file before
//  posting the notification, and results are read back from another
//  shared file after the Tweak processes the command.
//

import Foundation

/// Shared file paths (must match the Tweak's Tweak.x constants).
private let kGhostKitDir      = "/var/mobile/Library/GhostKit"
private let kCommandFilePath  = "/var/mobile/Library/GhostKit/command.plist"
private let kResultFilePath   = "/var/mobile/Library/GhostKit/result.plist"

/// Darwin notification name constants (must match the Tweak's Tweak.x constants).
public enum DarwinCommand: String {
    // Keychain
    case cleanKeychain         = "GhostKitCleanKeychain"
    case deepCleanKeychain     = "GhostKitDeepCleanKeychain"
    case deleteAllKeychains    = "GhostKitDeleteAllKeychains"
    case backupKeychain        = "GhostKitBackupKeychain"
    case restoreKeychain       = "GhostKitRestoreKeychain"
    case listKeychain          = "GhostKitListKeychain"

    // Identifier
    case refreshIDFA           = "GhostKitRefreshIDFA"
    case changeIdentifier      = "GhostKitChangeIdentifier"
    case getCurrentIdentifiers = "GhostKitGetCurrentIdentifiers"

    // Device reset
    case resetDevice           = "GhostKitResetDevice"
    case getDeviceInfo         = "GhostKitGetDeviceInfo"

    // Cache cleaner
    case cleanSystemResidue    = "GhostKitCleanSystemResidue"
    case cleanDatabaseCache    = "GhostKitCleanDatabaseCache"
    case cleanDataDirectory    = "GhostKitCleanDataDirectory"
    case cleanCookies          = "GhostKitCleanCookies"
    case cleanPasteboard       = "GhostKitCleanPasteboard"
    case getAppSize            = "GhostKitGetAppSize"

    // App list
    case getAllApps            = "GhostKitGetAllApps"
    case searchApps            = "GhostKitSearchApps"
    case uninstallApp          = "GhostKitUninstallApp"
    case getAppInfo            = "GhostKitGetAppInfo"

    // Account
    case getAccountList        = "GhostKitGetAccountList"
    case addAccount            = "GhostKitAddAccount"
    case deleteAccount         = "GhostKitDeleteAccount"
    case getCurrentAccount    = "GhostKitGetCurrentAccount"
    case switchAccount         = "GhostKitSwitchAccount"

    // Permission
    case allowPasteForAll      = "GhostKitAllowPasteForAll"
    case isPasteAllowed        = "GhostKitIsPasteAllowed"

    // System
    case safeExit              = "GhostKitSafeExit"
    case respring              = "GhostKitRespring"
    case ldrestart             = "GhostKitLdrestart"

    // Injection
    case injectDylib           = "GhostKitInjectDylib"
    case removeDylib           = "GhostKitRemoveDylib"
    case getInjectedDylibs     = "GhostKitGetInjectedDylibs"

    // Graphics
    case applyGraphicsConfig   = "GhostKitApplyGraphicsConfig"
    case getAvailablePresets   = "GhostKitGetAvailablePresets"
    case getCurrentGraphics    = "GhostKitGetCurrentGraphics"
    case restoreDefaultGraphics = "GhostKitRestoreDefaultGraphics"
}

public class DarwinNotifyManager {

    public static let shared = DarwinNotifyManager()

    private let fm = FileManager.default

    private init() {
        // Ensure the shared directory exists.
        try? fm.createDirectory(atPath: kGhostKitDir,
                                withIntermediateDirectories: true)
    }

    // MARK: - Public API

    /// Send a Darwin command to the Tweak layer with optional parameters.
    /// Blocks until the Tweak writes a result file or times out.
    ///
    /// - Parameters:
    ///   - command: The Darwin command to send.
    ///   - params: Optional parameters dictionary (e.g., ["bundleID": "com.example.app"]).
    ///   - timeout: Maximum seconds to wait for a result (default 10s).
    /// - Returns: The result dictionary from the Tweak, or nil on timeout.
    public func sendCommand(_ command: DarwinCommand,
                            params: [String: Any]? = nil,
                            timeout: TimeInterval = 10.0) -> [String: Any]? {

        // 1. Remove any stale result file.
        removeResultFile()

        // 2. Write command parameters to the shared file.
        var commandDict: [String: Any] = ["command": command.rawValue]
        if let params = params {
            commandDict["params"] = params
        }
        let success = (commandDict as NSDictionary).write(toFile: kCommandFilePath, atomically: true)
        guard success else {
            NSLog("[GhostKit] DarwinNotify: failed to write command file")
            return nil
        }

        // 3. Post the Darwin notification.
        let center = CFNotificationCenterGetDarwinNotifyCenter()
        let name = command.rawValue as CFString
        CFNotificationCenterPostNotification(center, CFNotificationName(name), nil, nil, true)

        // 4. Poll for the result file.
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            if let result = readResultFile() {
                // Verify this result is for our command.
                if let cmd = result["command"] as? String,
                   cmd == command.rawValue {
                    return result
                }
            }
            usleep(50_000)  // 50ms between polls
        }

        NSLog("[GhostKit] DarwinNotify: timeout waiting for result for %@", command.rawValue)
        return nil
    }

    /// Send a fire-and-forget Darwin command (no result expected).
    /// Used for commands like respring/ldrestart that may kill the process.
    public func sendFireAndForget(_ command: DarwinCommand,
                                 params: [String: Any]? = nil) {
        // Write command parameters.
        var commandDict: [String: Any] = ["command": command.rawValue]
        if let params = params {
            commandDict["params"] = params
        }
        (commandDict as NSDictionary).write(toFile: kCommandFilePath, atomically: true)

        // Post the Darwin notification.
        let center = CFNotificationCenterGetDarwinNotifyCenter()
        let name = command.rawValue as CFString
        CFNotificationCenterPostNotification(center, CFNotificationName(name), nil, nil, true)
    }

    // MARK: - Private helpers

    private func removeResultFile() {
        try? fm.removeItem(atPath: kResultFilePath)
    }

    private func readResultFile() -> [String: Any]? {
        guard fm.fileExists(atPath: kResultFilePath) else {
            return nil
        }
        return NSDictionary(contentsOfFile: kResultFilePath) as? [String: Any]
    }

    // MARK: - Convenience methods

    /// Send a command that operates on a specific bundle ID.
    public func sendBundleIDCommand(_ command: DarwinCommand,
                                    bundleID: String,
                                    timeout: TimeInterval = 10.0) -> Bool {
        let result = sendCommand(command, params: ["bundleID": bundleID], timeout: timeout)
        if let res = result?["result"] as? Bool {
            return res
        }
        if let res = result?["result"] as? Int {
            return res != 0
        }
        return false
    }

    /// Send a command that returns a dictionary result.
    public func sendDictionaryCommand(_ command: DarwinCommand,
                                      bundleID: String? = nil,
                                      timeout: TimeInterval = 10.0) -> [String: Any] {
        var params: [String: Any] = [:]
        if let bid = bundleID { params["bundleID"] = bid }
        let result = sendCommand(command, params: params.isEmpty ? nil : params, timeout: timeout)
        return (result?["result"] as? [String: Any]) ?? [:]
    }

    /// Send a command that returns an array result.
    public func sendArrayCommand(_ command: DarwinCommand,
                                  bundleID: String? = nil,
                                  timeout: TimeInterval = 10.0) -> [[String: Any]] {
        var params: [String: Any] = [:]
        if let bid = bundleID { params["bundleID"] = bid }
        let result = sendCommand(command, params: params.isEmpty ? nil : params, timeout: timeout)
        return (result?["result"] as? [[String: Any]]) ?? []
    }

    /// Send a command that returns a string result.
    public func sendStringCommand(_ command: DarwinCommand,
                                   bundleID: String? = nil,
                                   timeout: TimeInterval = 10.0) -> String {
        var params: [String: Any] = [:]
        if let bid = bundleID { params["bundleID"] = bid }
        let result = sendCommand(command, params: params.isEmpty ? nil : params, timeout: timeout)
        return (result?["result"] as? String) ?? ""
    }
}
